crest/Assets/Crest/Crest/Scripts/Collision/CollProviderBakedFFT.cs

```diff
@@ -0,0 +1,469 @@
+{
+    /// <summary>
+    /// Collision provider for baked FFT data
+    /// </summary>
+    public class CollProviderBakedFFT : ICollProvider
+    {
+        enum QueryStatus
+        {
+            Success,
+            DataMissing,
+        }
+
+        public FFTBakedData _data = null;
+
+        const float s_finiteDiffDx = 0.1f;
+        const float s_finiteDiffDt = 0.06f;
+        const int s_jobBatchSize = 8;
+
+        public CollProviderBakedFFT(FFTBakedData data)
+        {
+            Debug.Assert(data != null, "Crest: Baked data should not be null.");
+            _data = data;
+        }
+
+        public int Query(int i_ownerHash, float i_minSpatialLength, Vector3[] i_queryPoints, Vector3[] o_resultDisps, Vector3[] o_resultNorms, Vector3[] o_resultVels)
+        {
+            if (_data == null || _data._framesFlattenedNative.Length == 0)
+                return (int)QueryStatus.DataMissing;
+
+            var t = OceanRenderer.Instance.CurrentTime;
+
+            // Queries processed in groups of 4 for SIMD - 'quads'
+            var numQueryQuads = (o_resultDisps.Length + 3) / 4;
+            var queryPointsX = new NativeArray<float4>(numQueryQuads, Allocator.TempJob, NativeArrayOptions.ClearMemory);
+            var queryPointsZ = new NativeArray<float4>(numQueryQuads, Allocator.TempJob, NativeArrayOptions.ClearMemory);
+
+            // Copy input data. Could be avoided if query api is changed to use NAs.
+            for (int i = 0; i < i_queryPoints.Length; i++)
+            {
+                var quadIdx = i / 4;
+                var xQuad = queryPointsX[quadIdx];
+                var zQuad = queryPointsZ[quadIdx];
+
+                var quadComp = i % 4;
+                xQuad[quadComp] = i_queryPoints[i].x;
+                zQuad[quadComp] = i_queryPoints[i].z;
+
+                queryPointsX[quadIdx] = xQuad;
+                queryPointsZ[quadIdx] = zQuad;
+            }
+
+            if (o_resultDisps != null)
+            {
+                // One thread per quad - per group of 4 queries
+                var resultsX = new NativeArray<float4>(numQueryQuads, Allocator.TempJob, NativeArrayOptions.UninitializedMemory);
+                var resultsY = new NativeArray<float4>(numQueryQuads, Allocator.TempJob, NativeArrayOptions.UninitializedMemory);
+                var resultsZ = new NativeArray<float4>(numQueryQuads, Allocator.TempJob, NativeArrayOptions.UninitializedMemory);
+
+                // Run job synchronously
+                new JobSampleDisplacement
+                {
+                    _queryPointsX = queryPointsX,
+                    _queryPointsZ = queryPointsZ,
+                    _framesFlattened = _data._framesFlattenedNative,
+                    _t = t,
+                    _params = _data._parameters,
+                    _outputX = resultsX,
+                    _outputY = resultsY,
+                    _outputZ = resultsZ,
+                }.Schedule(numQueryQuads, s_jobBatchSize).Complete();
```

Comment:
It would probably be a good performance improvement if instead of `Complete`ing 3 jobs synchronously, they were combined (`CombineDependencies` I think?) into a single JobHandle which would be then `Complete`d and finally assign values to results. There's a lot of waiting done on job threads currently.

Reply:
this is done on latest, thanks